cdap-watchdog/src/main/java/co/cask/cdap/logging/write/LogCleanup.java

```diff
@@ -109,13 +127,191 @@ public Void call() throws Exception {
+                                                                                              logBaseDir, namespaceId,
+                                                                                              impersonator);
+      // processor to get list of log files older than retention duration
+      final Processor<LocationStatus, Set<Location>> diskFilesProcessor = getDiskFilesProcessor(namespaceId, tillTime);
+      if (namespacedLogBaseDir.exists()) {
+        // Get all the disk log files on disk for a given namespace
+        Set<Location> diskFileLocations = getDiskLocations(namespaceId, namespacedLogBaseDir, diskFilesProcessor);
+        // remove log files from diskFileLocations for which metadata is present.
+        removeLocationsWithMeta(tillTime, namespaceId, diskFileLocations);
+
+        // delete all the disk locations for which metadata is not present
+        for (final Location locationToDelete : diskFileLocations) {
+          try {
+            deleteLogFiles(parentDirs, namespaceId, namespacedLogBaseDir.toString(), locationToDelete);
+          } catch (Exception e) {
+            LOG.warn("Got exception when deleting path {}", locationToDelete, e);
+          }
+          namespacedLogBaseDirMap.put(namespacedLogBaseDir.toString(), namespaceId);
+        }
+      }
+    }
+  }
+
+  /**
+   * Remove all the disk files from diskFileLocations for which meta data is present
+   * @param tillTime time till the meta data will be deleted.
+   * @param namespaceId namespace for which metadata needs to be scanned
+   * @param diskFileLocations log files present on disk for given namespace
+   */
+  private void removeLocationsWithMeta(final long tillTime, NamespaceId namespaceId,
+                                       final Set<Location> diskFileLocations) {
+    // get the logging context for a given namespace
+    LoggingContext loggingContext = new LogNamespaceLoggingContext(namespaceId.getNamespace());
+    FileMetaDataManager.TableKey nextTableKey = new FileMetaDataManager.TableKey(loggingContext.getLogPartition(),
+                                                                                 null);
+    Processor<URI, Set<URI>> metaProcessor = getMetaFilesProcessor();
+    // Get all the log metadata in batches of 100
+    do {
+      nextTableKey = fileMetaDataManager.scanFiles(nextTableKey, MAX_META_FILES_SCANNED, metaProcessor);
+      // create location from scanned uris and remove all the metadata files modified before tillTime
+      for (final URI uri : metaProcessor.getResult()) {
+        try {
+          impersonator.doAs(namespaceId, new Callable<Void>() {
+            @Override
+            public Void call() throws Exception {
+              Location fileLocation = rootLocationFactory.create(uri);
+              if (fileLocation.lastModified() < tillTime && diskFileLocations.contains(fileLocation)) {
```

Comment:
We don't need to make impersonated file system call to get the last modified time for a file here, right? We can just remove all files that are in log meta from `diskFileLocations`. As long as a file is in log meta, it will get removed sooner or later.